src/main/ingestion-queues/kafka-queue.ts

```diff
@@ -51,39 +44,40 @@ export class KafkaQueue implements Queue {
         const eachEventStartTimer = new Date()
 
-        const processingTimeout = timeoutGuard('Still running plugins on event. Timeout warning after 30 sec!', {
-            event: JSON.stringify(event),
-        })
-        const timer = new Date()
-        let processedEvent: PluginEvent
-        try {
-            processedEvent = await this.processEvent(event)
-        } catch (error) {
-            status.info('🔔', error)
-            Sentry.captureException(error)
-            throw error
-        } finally {
-            this.pluginsServer.statsd?.timing('kafka_queue.single_event', timer)
-            clearTimeout(processingTimeout)
+        let processedEvent: PluginEvent | null = event
+
+        if (processedEvent.event !== '$snapshot') {
+            processedEvent = await this.runInstrumentedFunction({
+                event,
+                func: (event) => this.workerMethods.processEvent(event),
+                statsKey: 'kafka_queue.single_event',
+                timeoutMessage: 'Still running plugins on event. Timeout warning after 30 sec!',
+            })
         }
 
         // ingest event
-
         if (processedEvent) {
-            const singleIngestionTimeout = timeoutGuard('After 30 seconds still ingesting event', {
-                event: JSON.stringify(processedEvent),
-            })
-            const singleIngestionTimer = new Date()
-            try {
-                await this.ingestEvent(processedEvent)
-            } catch (error) {
-                status.info('🔔', error)
-                Sentry.captureException(error)
-                throw error
-            } finally {
-                this.pluginsServer.statsd?.timing('kafka_queue.single_ingestion', singleIngestionTimer)
-                clearTimeout(singleIngestionTimeout)
-            }
+            await Promise.all([
+                this.runInstrumentedFunction({
+                    event: processedEvent,
+                    func: (event) => this.workerMethods.ingestEvent(event),
+                    statsKey: 'kafka_queue.single_ingestion',
+                    timeoutMessage: 'After 30 seconds still ingesting event',
+                }),
+                processedEvent.event === '$snapshot'
+                    ? this.runInstrumentedFunction({
+                          event: processedEvent,
+                          func: (event) => this.workerMethods.onSnapshot(event),
+                          statsKey: 'kafka_queue.single_on_snapshot',
+                          timeoutMessage: 'After 30 seconds still running onSnapshot',
+                      })
+                    : this.runInstrumentedFunction({
+                          event: processedEvent,
+                          func: (event) => this.workerMethods.onEvent(event),
+                          statsKey: 'kafka_queue.single_on_event',
+                          timeoutMessage: 'After 30 seconds still running onEvent',
+                      }),
```

Comment:
Nit: I think it'd be a bit more readable to do the ternary operaration on arg _in_ the `this.runInstrumentedFunction(arg)` call.

Reply:
Indeed it was! I got carried away and refactored this ingestion dance into a separate `ingest-event.ts` simple [POJSF](https://gist.github.com/klaaspieter/12cd68f54bb71a3940eae5cdd4ea1764) as well, and shared it between Celery and Kafka. [Instrumentation must be free!](https://www.openbsd.org/lyrics.html#35)

Reply:
Nice, good refactor